core/src/impl/Kokkos_FunctorAnalysis.hpp

```diff
@@ -90,6 +90,29 @@ struct DeduceFunctorPatternInterface<ParallelScanWithTotal<
   using type = FunctorPatternInterface::SCAN;
 };
 
+#if defined(__CUDA_ARCH__)
+template <typename T>
+KOKKOS_INLINE_FUNCTION unsigned char volatile_preload(const volatile T* src,
+                                                      size_t count) {
+  // Ensure that subsequent code in this thread that may do
+  // non-`volatile` loads through `src` will not see stale
+  // non-coherent values in the cache by forcing `volatile` loads to
+  // precede it
+  auto ct         = reinterpret_cast<const volatile char*>(src);
+  unsigned char c = 0;
+  for (size_t i = 0; i < count * sizeof(T); ++i) {
```

Comment:
Given that Nvidia documents a memory transaction size of at least 32 bytes (or 128 bytes, depending on cache level servicing a request), we could potentially optimize this to `i += 32` to reduce the number of discrete instructions issued.

Reply:
Per documentation in this section: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#compute-capability-3-0

Reply:
Curiously, with our current tests, run on a V100, even incrementing by 4,096,000 still passes, though I really think it shouldn't. I guess our array reduction tests are not sensitive enough, relative to how the code ends up executing.

Reply:
... The test are apparently very insensitive in a `Release` build - commenting out the calls to `volatile_preload` entirely, they still pass!

Reply:
It's possible that the earlier changes in #4855 do effectively obviate the need for the `volatile` loads, though it's not entirely clear that they should